show Passthrough cover on track overview

ronso0 opened the pull request:
![image](https://user-images.githubusercontent.com/5934199/78095467-5afe4e80-73d7-11ea-939d-8418222374bc.png)

Inspired by #2575 

Show passthrough cover over entire overview, picks color from `<PassthroughOverlayColor>` node.
Stack _Passthrough_ label on top, accepts custom styles via qss: `WOverview #PassthroughLabel`.
All mouse events are still passed to the overview.

I added the `PassthroughOverlayColor` picker to /waveform/renderers/waveformsignalcolors.cpp, I guess it can also be used for the scrolling waveform then?

- [x] add styles to all skins
- [x] add default color to C++?

ronso0 commented:
It would be nice if someone colud review this to let me know if I'm on the right track.
If so, I'll continue to add the skin styles and a default color.

Holzhaus reviewed (commented):
LGTM so far, thanks!

ronso0 commented:
well, that was quick!

ronso0 commented:
> Be-ing removed the beta non-blocker label

why? "passthrough" is displayed over the overview #2575 , this upgrade just keeps the overview usable and can be added during beta (if I don't finish the skin implementation before that)

Be-ing commented:
"beta non-blocker" is a useless tag. Everything not tagged as a beta blocker is a non-blocker; having a tag for that is superfluous.

ronso0 commented:
There where some embarassing copy/paste mistakes and a bug with skin reload that slipped throught earlier, so I started over and rebased. Please review.
Some functions are also more effective now, for example the cover color is only read once during setup. I'll use that for the playedOverlayColor, too.

With default.qss it looks like this:
![image](https://user-images.githubusercontent.com/5934199/78512448-d29bf700-77a4-11ea-9b0f-d1412950963f.png)

Custom skin styles coming soon..

Be-ing commented on src/widget/woverview.cpp:
```diff
@@ -94,6 +93,17 @@ WOverview::WOverview(
+    m_pPassthroughLabel = new QLabel(this);
+    m_pPassthroughLabel->setObjectName("PassthroughLabel");
+    m_pPassthroughLabel->setAlignment(Qt::AlignLeft|Qt::AlignVCenter);
+    m_pPassthroughLabel->setText("Passthrough");
```
```suggestion
    //: Shown on the overview waveform when vinyl passthrough is enabled
    m_pPassthroughLabel->setText(tr("Passthrough"));
```

ronso0 replied:
okay, thx!

ronso0 commented:
![image](https://user-images.githubusercontent.com/5934199/78512967-c154e980-77a8-11ea-9204-876b7bd6986c.png)
![image](https://user-images.githubusercontent.com/5934199/78513028-11cc4700-77a9-11ea-9353-a9beba74a2df.png)

![image](https://user-images.githubusercontent.com/5934199/78513084-72f41a80-77a9-11ea-9e59-5a9bd7b37588.png)
![image](https://user-images.githubusercontent.com/5934199/78513116-afc01180-77a9-11ea-9433-bffe89f8cec5.png)
![image](https://user-images.githubusercontent.com/5934199/78513138-db42fc00-77a9-11ea-9359-31f40e9df4b3.png)

Tango with #2619 : no passthrough cover on Play, no highlight on Vinyl Controls toggle
![image](https://user-images.githubusercontent.com/5934199/78513310-e9454c80-77aa-11ea-8324-c7eb665ec90c.png)
![image](https://user-images.githubusercontent.com/5934199/78513322-fd894980-77aa-11ea-8890-cd40da719051.png)

![image](https://user-images.githubusercontent.com/5934199/78513514-94a2d100-77ac-11ea-87ca-08db3c953d13.png)

Holzhaus commented on src/widget/woverview.h:
```diff
@@ -170,6 +171,9 @@ class WOverview : public WWidget, public TrackDropTarget {
     QColor m_endOfTrackColor;
+    QColor m_passthroughOverlayColor;
+    QLabel* m_pPassthroughLabel;
+    QVBoxLayout* m_pPassThroughLayout;
```
`setLayout` transfers ownership to the widget, you don't need to save a reference to the layout.

ronso0 replied:
so `QVBoxLayout passThroughLayout = new ...QVBoxLayout()` would suffice?

Holzhaus replied:
```c++
QVBoxLayout* pPassThroughLayout = new QVBoxLayout();
```

ronso0 replied:
a got it..

Holzhaus commented on src/widget/woverview.cpp:
```diff
@@ -94,6 +93,18 @@ WOverview::WOverview(
+
+    m_pPassthroughLabel = new QLabel(this);
+    m_pPassthroughLabel->setObjectName("PassthroughLabel");
+    m_pPassthroughLabel->setAlignment(Qt::AlignLeft|Qt::AlignVCenter);
```
```suggestion
    m_pPassthroughLabel->setAlignment(Qt::AlignLeft | Qt::AlignVCenter);
```

ronso0 replied:
ok

Holzhaus reviewed (changes requested):
Great! Code looks good.

Now for the obligatory bikeshedding:
I think we should try to re-use the skin's track title label text colors for the Passthrough label. Different colors look a bit out of place. For shade, this would be dark grey instead of green, and in Tango blue or Fuchsia depending on color scheme. I'm not sure about Deere, but I think either white or blue (the same shade that is used for buttons) would look good.

ronso0 commented:
> Now for the obligatory bikeshedding:
> I think we should try to re-use the skin's track title label text colors for the Passthrough label. Different colors look a bit out of place.

Here we go :)
I think we should emphasize the _indicator_ nature of the Passthrough label because it's a different information level than a _track title_, so I chose scheme-specific indiciator/button colors*. IMO that helps to notice it, especially since the Passthrough toggles are either very small (Shade), can be hidden (Tango, Deere) or might be far away from the decks (Deere w/ stacked waveforms).
*only in LateNight this doesn't make difference

Holzhaus commented:
I converted this to draft to prevent premature merging because there are still things left to do (according to the checkboxes in your PR description). If this is ready, feel free to convert it back.

Holzhaus commented on src/widget/woverview.cpp:
```diff
@@ -94,6 +93,18 @@ WOverview::WOverview(
+    // Shown on the overview waveform when vinyl passthrough is enabled
+    m_pPassthroughLabel->setText(tr("Passthrough"));
+    m_pPassthroughLabel->hide();
+    QVBoxLayout *passThroughLayout = new QVBoxLayout(this);
```
```suggestion
    QVBoxLayout *pPassthroughLayout = new QVBoxLayout(this);
```

Holzhaus reviewed (approved):
Code LGTM so far (except for one variable name). Thank you.

ronso0 commented:
Thx, comments addressed, ToDos finished.
Ready for merge if the colors are reasonable.

ronso0 commented:
> I converted this to draft

Ha! Now I found that 'button', too.

Holzhaus reviewed (approved):
> Ready for merge if the colors are reasonable.

I still don't like the colors (except for LateNight), but this subjective and I don't want to start endless discussions about this. I you think they fit in nicely, so be it (I'm using LateNight anyway :wink:).

Holzhaus commented on src/widget/woverview.cpp:
```diff
@@ -94,6 +93,18 @@ WOverview::WOverview(
+    m_pPassthroughLabel->setText(tr("Passthrough"));
+    m_pPassthroughLabel->hide();
+    QVBoxLayout *pPassthroughLayout = new QVBoxLayout(this);
+    passThroughLayout->setContentsMargins(0,0,0,0);
```
You only changed the first occurrence.

```suggestion
    pPassthroughLayout->setContentsMargins(0,0,0,0);
```
and below

Holzhaus replied:
Please fixup the broken commits in this branch and force-push to allow us to git bisect later.

ronso0 replied:
Thanks for checking! Done.

ronso0 committed:
draw passthrough overlay on overview waveform

When vinyl passthrough is enabled, dimm waveform overviews with a
translucent overlay on top of all overview itemsgraphics and put
'Passthrough' QLabel on top.
Any mouse events are still passed to the overview widget.
Skin style:
The cover color is picked from skin node <PassthroughOverlayColor>
in the Overview widget (default #bb000000), the default label style
can be adjusted in qss (#PassthroughLabel).

ronso0 committed:
passthrough cover: add default & skin styles

ronso0 commented:
Fixed and squashed.
Ready for merge.

Holzhaus reviewed (commented):
Thanks!

 I retested. The font for Tango is incorrect on my machine:

![2020-04-15-105714_747x385_scrot](https://user-images.githubusercontent.com/1834516/79319000-0f37c300-7f08-11ea-96e0-ad753113ec8e.png)

Other than that, LGTM.

ronso0 committed:
Tango: remove Passthrough hint from VC toggle & Play button

ronso0 commented:
> I retested. The font for Tango is incorrect on my machine:

Strange, here it looks as intended.
I added `font-weight: bold;` to all stylesheets to ensure it on all OS.
After further testing I made the label color in Shade Dark a bit brighter.
For the sake of completeness I also removed the Passthrough hints from Tango. (might create conflicts for #2619 though)
Ready!

Holzhaus reviewed (commented):
Still looks like this:

![2020-04-15-115852_664x293_scrot](https://user-images.githubusercontent.com/1834516/79324889-90935380-7f10-11ea-9bfc-bae9184fb518.png)

ronso0 commented:
This is deck1 with ClubTwist, right?
I have no idea why the font style is not applied, ClubTwist doesn't have a separate stylesheet.
Issue should be solved after merging #2647

Holzhaus commented:
> This is deck1 with ClubTwist, right?
> I have no idea why the font style is not applied, ClubTwist doesn't have a separate stylesheet.
> Issue should be solved after merging #2647

I merged #2647 locally and the issue is indeed fixed.

Holzhaus reviewed (approved):
LGTM.
